src/main/java/com/algorand/algosdk/transaction/AtomicTransactionComposer.java

```diff
@@ -0,0 +1,449 @@
+            }
+            Response<PendingTransactionResponse> resp =
+                    client.PendingTransactionInformation(this.transactionList.get(i).txn.txID()).execute();
+            if (!resp.isSuccessful())
+                throw new IllegalArgumentException("cannot get response from client");
+
+            // question why it is not replying anything in log?
+            PendingTransactionResponse respBody = resp.body();
+            List<byte[]> logs = respBody.logs;
+            byte[] retLine = null;
+            for (int logIndex = logs.size() - 1; logIndex >= 0; logIndex--) {
+                byte[] line = logs.get(logIndex);
+                if (!checkLogRet(line))
+                    continue;
+                retLine = line;
+                break;
+            }
+            if (retLine == null)
+                throw new IllegalArgumentException("expected to find logged return value, got none");
+
+            byte[] abiEncoded = Arrays.copyOfRange(retLine, 4, retLine.length);
+            ABIValue decoded = null;
+            Exception parseError = null;
+            try {
+                ABIType ABIType = com.algorand.algosdk.abi.ABIType.Of(this.methodMap.get(i).returns.type);
+                Object decodedVar = ABIType.decode(abiEncoded);
+                decoded = new ABIValue(ABIType, decodedVar);
+            } catch (Exception e) {
+                parseError = e;
+            }
+            retList.add(new ReturnValue(
+                    this.transactionList.get(i).txn.txID(),
+                    abiEncoded,
+                    decoded,
+                    parseError
+            ));
+        }
+
+        this.status = AtomicTxComposerStatus.COMMITTED;
+
+        return new ExecuteResult(txInfo.confirmedRound, this.getTxIDs(), retList);
+    }
+
+    private static boolean checkLogRet(byte[] logLine) {
+        if (logLine.length < 4)
+            return false;
+        for (int i = 0; i < 4; i++) {
+            if (logLine[i] != ABI_RET_HASH[i])
+                return false;
+        }
+        return true;
+    }
+
+    public static class ExecuteResult {
+        public Long confirmedRound;
+        public List<String> txIDs;
+        public List<ReturnValue> methodResults;
+
+        public ExecuteResult(Long confirmedRound, List<String> txIDs, List<ReturnValue> methodResults) {
+            this.confirmedRound = confirmedRound;
+            this.txIDs = txIDs;
+            this.methodResults = methodResults;
+        }
+    }
+
+    public static class ReturnValue {
+        public String txID;
+        public byte[] rawValue;
+        public ABIValue value;
+        public Exception parseError;
+
+        public ReturnValue(String txID, byte[] rawValue, ABIValue value, Exception parseError) {
+            this.txID = txID;
+            this.rawValue = rawValue;
+            this.value = value;
+            this.parseError = parseError;
+        }
+    }
+
+    public static class TransactionSigner {
```

Comment:
This would be a lot better as interface like I suggested here, but you may not have seen it: https://github.com/algorand/java-algorand-sdk/issues/264#issuecomment-942584208

The reason being we likely want to add more types of transaction signers in the future for things like KMD or WalletConnect/other wallet protocols. For this reason it's also probably better if TransactionSigner were declared in its own file instead of in this one.

Additionally if TransactionSigner were an interface, you could create anonymous implementations like this:

```java
    // inside of src/main/java/com/algorand/algosdk/account/Account.java
    public TransactionSigner getTransactionSigner() {
        Account self = this;
        return new TransactionSigner() {
            @Override
            public SignedTransaction[] signTransactions(Transaction[] txnGroup, int[] indexesToSign) throws Exception {
                SignedTransaction[] stxns = new SignedTransaction[indexesToSign.length];
                for (int i = 0; i < indexesToSign.length; i++) {
                    stxns[i] = self.signTransaction(txnGroup[indexesToSign[i]]);
                }
                return stxns;
            }
        };
    }
```

Reply:
That is way better than my previous thoughts. I was constantly wondering how to get that integrate with other stuffs like KMD or wallet protocols, etc. This one seems more reasonable.

Reply:
But another question arise: what if we want to use multi-sig, do we need to provide all the secret key and generate the `TransactionSigner`?

Reply:
Yes a multisig TransactionSigner will need a list of private keys to sign with. In the future we might want to revisit this to be more modular (maybe you can "join" basic TransactionSigners backed by a single private key into one multisig signer), but for simplicity I think this is the best approach for now

Reply:
A follow-up question. I know we want to have *only one* call of sign from each `TransactionSigner`. Thus we might want to distinguish each interface value from another.

I think in this case we want to expose a method called `getAddress()`, which returns an address from each signer. In this way we can build a map of `TransactionSigner -> int[]`, which allows you to sign transaction group once from the signer.

Reply:
That mostly sounds good, except I think it would be more general to use the `equals()` method to see if two `TransactionSigner`s are equivalent. The reason being there are some edge cases where two different signers need to sign for the same address (e.g. a delegated LogicSig can sign transactions for an account, but so can that account's secret key).

You could add the `equals` method as an abstract method on `TransactionSigner`, but then every instance would have to implement it. And currently since you are using anonymous instances, this would be hard to do.

So I'd probably suggest calling `equals()` from the `AtomicTransactionComposer` to check if `TransactionSigner`s are equivalent, but not actually implementing the `equals()` method on any instances. This results in Java comparing the objects by reference, which is fine for now, and something we can always improve later.

Reply:
This sounds mostly reasonable to me, but I am concerned in the last part, since it is not necessarily true to have same instances of signer. I am wondering if I can use extension of interface like `comparable`, which might make things easier?